### PR TITLE
nodeport: fix CT accounting in revDNAT path

### PR DESCRIPTION
**Pull request by julianwiedmann**
While working on https://github.com/cilium/cilium/issues/22659, I noticed that we mis-account RX traffic in the nodeport revDNAT path. Background is that we first do a CT lookup for a packet (and update the statistics while at it), but then decide that the matched CT entry doesn't fit the criteria for nodeport revDNAT.

```release-note
Fix double-accounted RX packets in CT statistics when Nodeport is in use.
```

### CODE REVIEW

**Review by YutaroHayakawa — APPROVED** [review]
LGTM, thanks!

**Comment by julianwiedmann** [comment]
`test-1.25-4.19` hit some rate-limiting in docker.io, re-running ...

**Comment by julianwiedmann** [comment]
/test-1.25-4.19

File: bpf/lib/nodeport.h [hunk 1]
```diff
@@ -2005,6 +2006,18 @@ static __always_inline int rev_nodeport_lb4(struct __ctx_buff *ctx, __u32 *ifind
...
+#ifdef ENABLE_MASQUERADE
+	ret = ct_extract_ports4(ctx, l4_off, CT_EGRESS, &tuple);
+	if (ret == DROP_CT_UNKNOWN_PROTO)
+		goto skip_nat;
+	if (ret < 0)
+		return ret;
+
+	if (!ct_has_nodeport_egress_entry4(get_ct_map4(&tuple), &tuple))
+		goto skip_nat;
+#endif /* ENABLE_MASQUERADE */
```
Comment (julianwiedmann) [inline bpf/lib/nodeport.h]: Thinking we might actually need to match the full condition in https://github.com/cilium/cilium/blob/4ddeb34a9192a3634d816a915c12629cac5340b7/bpf/bpf_host.c#L1196_L1201. And thus also add a similar check in the IPv6 path.

And bpf_overlay is even more [inclusive](https://github.com/cilium/cilium/blob/4ddeb34a9192a3634d816a915c12629cac5340b7/bpf/bpf_overlay.c#L563) in its SNAT usage :thinking: .

Comment (julianwiedmann) [reply bpf/lib/nodeport.h]: > Thinking we might actually need to match the full condition in https://github.com/cilium/cilium/blob/4ddeb34a9192a3634d816a915c12629cac5340b7/bpf/bpf_host.c#L1196_L1201. And thus also add a similar check in the IPv6 path.
> 
Yeah, just checked. The CT entries for hostns-originating connections (which we track due to potential SNAT conflicts vs forwarded Nodeport connections) get the same RX double accounting. Will update.

File: bpf/lib/nodeport.h [hunk 2]
```diff
@@ -1110,15 +1110,16 @@ static __always_inline int rev_nodeport_lb6(struct __ctx_buff *ctx, __u32 *ifind
...
-			ep_tail_call(ctx, CILIUM_CALL_IPV6_FROM_NETDEV);
-			return DROP_MISSED_TAIL_CALL;
-		}
+
+		return CTX_ACT_REDIRECT;
 	}
 
-	return CTX_ACT_REDIRECT;
+	if (bpf_skip_recirculation(ctx))
+		return DROP_NAT_NO_MAPPING;
```
Comment (borkmann) [inline bpf/lib/nodeport.h:1104]: Makes sense to drop given the ifindex 0.. so there's no known call-site where this was set?

Comment (julianwiedmann) [reply bpf/lib/nodeport.h:1104]: > Makes sense to drop given the ifindex 0.. so there's no known call-site where this was set?

Not sure I follow. The `from-container` path sets the `skip-recirculation` flag before it tail-calls, and we **should** always find a matching CT entry. So this just handles the "impossible" condition that we don't find a CT entry.

**Comment by julianwiedmann** [comment]
Quick respin:

- Rebase on-top of @borkmann 's Stateful GW fixes
- Widen the scope. We need to consider more than just BPF-Masquerade, host-originating connections are affected as well.
- I tried to compensate for the additional overhead in the RevDNAT path, by only calling into that code when we "really" need it. There's probably some room for more optimization there, but think it's good enough for now...

**Review by YutaroHayakawa — APPROVED** [review]

File: bpf/lib/nodeport.h [hunk 3]
```diff
@@ -1091,15 +1091,16 @@ static __always_inline int rev_nodeport_lb6(struct __ctx_buff *ctx, __u32 *ifind
...
```
Comment (pchaigno) [inline bpf/lib/nodeport.h:1104]: If I understand correctly the impact of this commit, in case of a mismatch between `egress_ct_state->node_port` and `ingress_ct_state` (i.e., first is true but second doesn't exist for some reason), we would silently drop the packet because we would call `bpf_redirect` with an incorrect ifindex. Now, we will emit a Cilium drop event.

So clearly a good change. And not a bugfix because we don't expect this mismatch to ever happen today (correct me if wrong @borkmann).

Nevertheless, it is a change in behavior, so could we explicitly mention it in the commit message? That is, mention the user-visible impact.

File: bpf/lib/nodeport.h [hunk 4]
```diff
@@ -1640,14 +1644,22 @@ int tail_nodeport_nat_ingress_ipv4(struct __ctx_buff *ctx)
 
 	/* At this point we know that a reverse SNAT mapping exists.
 	 * Otherwise, we would have tail-called back to
-	 * CALL_IPV4_FROM_NETDEV in the code above. The existence of the
-	 * mapping is an indicator that the packet might be a reply from
-	 * a remote backend. So handle the service reverse DNAT (if
-	 * needed)
+	 * CALL_IPV4_FROM_NETDEV in the code above.
+	 */
+#if !defined(ENABLE_DSR) || (defined(ENABLE_DSR) && defined(ENABLE_DSR_HYBRID)) ||	\
```
Comment (pchaigno) [inline bpf/lib/nodeport.h:1650]: I think it would be worth having a code comment here to mention why we have a condition on DSR all of a sudden. IIUC, something like "Reverse DNAT for DSR was performed at the backend's node so we can skip that case for the intermediate node here."

File: bpf/lib/nodeport.h [hunk 5]
```diff
@@ -1640,14 +1644,22 @@ int tail_nodeport_nat_ingress_ipv4(struct __ctx_buff *ctx)
...
+    (defined(ENABLE_EGRESS_GATEWAY) && !defined(TUNNEL_MODE))
```
Comment (pchaigno) [inline bpf/lib/nodeport.h:1651]: Could you expand on this case in a code comment or in the commit? It seems like the trickiest case, yet it's relegated to a footnote :sweat_smile:  
In particular, why does `TUNNEL_MODE` make a difference here? Doesn't egress gateway traffic go through the tunnel regardless of the routing mode?

File: bpf/lib/nodeport.h [hunk 6]
```diff
@@ -1995,11 +1992,12 @@ static __always_inline int rev_nodeport_lb4(struct __ctx_buff *ctx, __u32 *ifind
...
 #endif /* ENABLE_EGRESS_GATEWAY */
 
-	tuple.nexthdr = ip4->protocol;
-	tuple.daddr = ip4->daddr;
-	tuple.saddr = ip4->saddr;
+	ret = ipv4_extract_tuple(ctx, &tuple, &l4_off);
+	if (ret == DROP_CT_UNKNOWN_PROTO)
+		goto out;
+	if (IS_ERR(ret))
+		return ret;
```
Comment (pchaigno) [inline bpf/lib/nodeport.h:2005]: What's the rationale for this change? If it's just a performance optimization, note that it likely increases verifier complexity; I'm not sure the tradeoff is worth it.

Do we expect any user impact or is it just supposed to bypass useless code? The `IS_ERR(ret)` case in particular makes me think there might be some user impact.

File: bpf/lib/nodeport.h [hunk 7]
```diff
@@ -992,21 +996,23 @@ static __always_inline int rev_nodeport_lb6(struct __ctx_buff *ctx, __u32 *ifind
...
 	if (nat_46x64_fib)
 		goto skip_rev_dnat;
 #endif
-	l4_off = l3_off + hdrlen;
+
+	if (l4_load_ports(ctx, l4_off, &tuple.dport) < 0)
+		return DROP_CT_INVALID_HDR;
+
+	if (!ct_has_nodeport_egress_entry6(get_ct_map6(&tuple), &tuple))
+		goto out;
```
Comment (pchaigno) [inline bpf/lib/nodeport.h:1010]: IIUC, only this last commit is a bugfix, so only this commit should be backported to stable branches. Unless I'm missing some dependency between the commits (?), the four previous commits can go into another pull request and don't need to be backported.

**Review by pchaigno — CHANGES_REQUESTED** [review]
Tricky stuff. Thanks for looking into this logic :muscle: It's definitely needed.

Couple comments below. Most are to make sure I understood and ask for more details in commits/code comments. The last one is to discuss what we should actually backport.

File: bpf/lib/nodeport.h (hunk 5 above)
Comment (julianwiedmann) [reply bpf/lib/nodeport.h:1651]: Just to be clear - you saw the comments right below this line, but they're not sufficient?

File: bpf/lib/nodeport.h (hunk 3 above)
Comment (julianwiedmann) [reply bpf/lib/nodeport.h:1104]: > Nevertheless, it is a change in behavior, so could we explicitly mention it in the commit message?

At this stage I don't expect it to have any user-visible impact (ie. it's totally an "impossible to hit" condition). But we do need it in the following patches to have a clean `goto out` path.

File: bpf/lib/nodeport.h (hunk 6 above)
Comment (julianwiedmann) [reply bpf/lib/nodeport.h:2005]: We shouldn't try to extract the L4 ports from a packet with unknown / unsupported L4 protocol.

File: bpf/lib/nodeport.h (hunk 7 above)
Comment (julianwiedmann) [reply bpf/lib/nodeport.h:1010]: We could arguably defer patches 2/3. But I believe they're worth having in this series to clarify the current behavior, and self-document what sort of traffic is actually affected by the subsequent changes in the RevDNAT code.

ie. if we missed any condition in the "this needs to continue in `CILIUM_CALL_IPV4_NODEPORT_REVNAT`" check, that's an indication that the RevDNAT changes themselves might also be incomplete.

File: bpf/lib/nodeport.h (hunk 5 above)
Comment (pchaigno) [reply bpf/lib/nodeport.h:1651]: Yes I did. The only comment below on the egress gateway case describes what is being done, not why:

> Also redirect EgressGW reply traffic into tunnel.

File: bpf/lib/nodeport.h (hunk 6 above)
Comment (pchaigno) [reply bpf/lib/nodeport.h:2005]: But we already check for the unknown protocol cases (i.e., not UDP, TCP, SCTP, or ICMP) in `ct_lookup{4,6}`. So why do we need to duplicate the check here?

Comment (julianwiedmann) [reply bpf/lib/nodeport.h:2005]: Ah, I see what you mean. In patch 4 we could/should tolerate the `DROP_CT_UNKNOWN_PROTO` and just continue. But for patch 5 we then strictly need that check - `ct_has_nodeport_egress_entry4()` expects the full 5-tuple, so we need to manually load the ports before even calling into `ct_lookup()`.

[long-term we can add an optimized ct_lookup_with_tuple() that doesn't extract the ports a second time. I suspect that would apply in other places too]

File: bpf/lib/nodeport.h (hunk 5 above)
Comment (julianwiedmann) [reply bpf/lib/nodeport.h:1651]: Ah - yes, we're piggy-backing on the RevDNAT code for EgressGW replies. The details are in https://github.com/cilium/cilium/blob/1b4f2a334a8f3826c925a5bf49b7fc1165f292f4/bpf/lib/nodeport.h#L1976-L1981

Yes, that's not pretty - and we should probably do a s/CILIUM_CALL_IPV4_NODEPORT_REVNAT/CILIUM_CALL_IPV4_NODEPORT_NAT_INGRESS_CONTINUE to clarify that it's also handling some non-RevDNAT cases.

File: bpf/lib/nodeport.h (hunk 6 above)
Comment (pchaigno) [reply bpf/lib/nodeport.h:2005]: Ah, I see. That makes sense. Could we have a small code comment to explain that?

File: bpf/lib/nodeport.h (hunk 5 above)
Comment (pchaigno) [reply bpf/lib/nodeport.h:1651]: Ouch! I see. We probably need a code comment here as well; it may be enough to point to "the egress gateway special case in rev_nodeport_lb{4,6}".

Comment (julianwiedmann) [reply bpf/lib/nodeport.h:1651]: Added a comment:

> Also let rev_nodeport_lb4() redirect EgressGW reply traffic into tunnel (see there for details).

File: bpf/lib/nodeport.h (hunk 4 above)
Comment (julianwiedmann) [reply bpf/lib/nodeport.h:1650]: Added a comment

> If we're not in full DSR mode, reply traffic from remote backends might pass back through the LB node and requires revDNAT.

File: bpf/lib/nodeport.h (hunk 6 above)
Comment (julianwiedmann) [reply bpf/lib/nodeport.h:2005]: I ended up squashing those two patches together, and using a new LB helper that handles the port extraction internally.

**Comment by julianwiedmann** [comment]
> Tricky stuff. Thanks for looking into this logic muscle It's definitely needed.
> 
> Couple comments below. Most are to make sure I understood and ask for more details in commits/code comments. The last one is to discuss what we should actually backport.

Needed one more rebase, and incorporated the review feedback. Please have another look :).

**Review by pchaigno — APPROVED** [review]

**Comment by julianwiedmann** [comment]
> do we still want to backport this? it was dropped from backports PRs and has been merged for 4 months

Thanks for the ping! Nope, don't think that's going to happen :disappointed:. I removed the label from this and the related PRs.
